### PR TITLE
RNG unit tests [WIP]

### PR DESCRIPTION
**Pull request by miterosan**
Needs feedback from @peppy

### CODE REVIEW

File: osu.Framework/MathUtils/RandomProvider.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+
+namespace osu.Framework.MathUtils
+{
+    internal class RandomProvider : IRandomProvider
+    {
+
+        public Random _random;
```
Comment (smoogipooo) [inline osu.Framework/MathUtils/RandomProvider.cs]: This, again, I don't like. Means somewhere in the code someone may do something dodgy like `((RandomProvider)RNG.Random)._random = ...`.

File: osu.Framework.Tests/MathUtils/TestRNG.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    [TestFixture]
+    class TestRNG
+    {
+        [Test]
+        public void TestNext()
+        {
+            Mock<IRandomProvider> mockRandomProvider = new Mock<IRandomProvider>();
+            int expectedRandomNumber = 134123;
+            mockRandomProvider.Setup((mock) => mock.Next()).Returns(expectedRandomNumber);
+            RNG.Random = mockRandomProvider.Object;
```
Comment (smoogipooo) [inline osu.Framework.Tests/MathUtils/TestRNG.cs]: What's the point? Why not just create a class if you do this? Why not just `Assert.AreEqual(RNG.Next, 134123)`?

File: osu.Framework.Tests/MathUtils/TestRNG.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+            int actualRandomNumber = RNG.Next();
```
Comment (smoogipooo) [inline osu.Framework.Tests/MathUtils/TestRNG.cs]: This is going to be very run-order specific and must be avoided.

Comment (smoogipooo) [reply osu.Framework.Tests/MathUtils/TestRNG.cs]: Actually just noticed that this is basically just doing a straight up `Assert.AreEqual(134123, 134123)` due to the redefinition of RNG.Random above, at which point this test case becomes useless.

Might as well call this a test case for Moq at this point...

Comment (peppy) [reply osu.Framework.Tests/MathUtils/TestRNG.cs]: this is how black-box testing works. you don't know the implementation of our `Random` class and only care about the inputs and outputs. the real issue here is that the class being tested is a very simple wrapper around Random to *make* it static.

a pretty redundant class to test, but the testing methodology here isn't incorrect. if we want 100% test coverage we do need tests like this

Comment (smoogipooo) [reply osu.Framework.Tests/MathUtils/TestRNG.cs]: The methodology is correct, but we should not be making Random public here for the purpose of allowing it as input in black box testing.

File: osu.Framework/MathUtils/RandomProvider.cs (hunk 1 above)
Comment (miterosan) [reply osu.Framework/MathUtils/RandomProvider.cs]: public was by accident. Not intended!

File: osu.Framework/MathUtils/RNG.cs [hunk 4]
```diff
@@ -11,34 +11,34 @@ namespace osu.Framework.MathUtils
     public static class RNG
     {
         // Base RNG. Maybe expose methods for re-seeding in the future?
-        private static readonly Random random = new Random();
+        static public IRandomProvider Random { private get; set; } = (IRandomProvider) new RandomProvider();
```
[the remark this answers is not in the document]
Comment (miterosan) [reply osu.Framework/MathUtils/RNG.cs]: Yes, you are right, *but* I didn't want to break any existing code by adding unit tests!

File: osu.Framework.Tests/MathUtils/TestRNG.cs (hunk 2 above)
Comment (miterosan) [reply osu.Framework.Tests/MathUtils/TestRNG.cs]: Seperating the three phases of testing.
Preparing
Executing 
Validating

File: osu.Framework/MathUtils/RNG.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (miterosan) [reply osu.Framework/MathUtils/RNG.cs]: So what do we do here now? Breaking master by changing how RNG works is not really my intention.

File: osu.Framework.Tests/MathUtils/TestRNG.cs (hunk 3 above)
Comment (miterosan) [reply osu.Framework.Tests/MathUtils/TestRNG.cs]: Whats about marking RNG.Random internal? And then using [Friend Assemblies](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/assemblies-gac/friend-assemblies)

**Pull request closed without merging**